integrations/outlook/src/index.ts

```diff
@@ -0,0 +1,252 @@
+        },
+        location: async () => {
+          throw new NotImplementedError()
+        },
+        carousel: async () => {
+          throw new NotImplementedError()
+        },
+        card: async () => {
+          throw new NotImplementedError()
+        },
+        choice: async (props) => {
+          const graphClient = getClient(props.ctx.configuration)
+          let content = `${props.payload.text}\n`
+
+          for (const option of props.payload.options) {
+            content += `- ${option.label}\n`
+          }
+          content += `\n\n${props.ctx.configuration.emailSignature}`
+          await graphClient.sendMail({
+            ...props,
+            body: {
+              contentType: 'text',
+              content,
+            },
+          })
+        },
+        dropdown: async (props) => {
+          const graphClient = getClient(props.ctx.configuration)
+          let content = `${props.payload.text}\n`
+
+          for (const option of props.payload.options) {
+            content += `- ${option.label}\n`
+          }
+          content += `\n\n${props.ctx.configuration.emailSignature}`
+          await graphClient.sendMail({
+            ...props,
+            body: {
+              contentType: 'text',
+              content,
+            },
+          })
+        },
+        html: async (props) => {
+          const graphClient = getClient(props.ctx.configuration)
+          await graphClient.sendMail({
+            ...props,
+            body: {
+              contentType: 'html',
+              content: props.payload.content,
+            },
+          })
+        },
+      },
+    },
+  },
+  handler: async ({ req, ctx, client }) => {
+    // If there is a validationToken parameter
+    // in the query string, this is the endpoint validation
+    // request sent by Microsoft Graph. Return the token
+    // as plain text with a 200 response
+    // https://docs.microsoft.com/graph/webhooks#notification-endpoint-validation
+    if (req.query) {
+      const queryString = req.query as string
+      const queryParameters = queryString.split('&')
+      const queryParams: { [key: string]: string } = {}
+
+      for (const param of queryParameters) {
+        const [key, value] = param.split('=')
+        if (key && value) {
+          queryParams[key] = value
+        }
+      }
+
+      const validationToken = queryParams.validationToken
+
+      if (validationToken) {
+        var str = decodeURIComponent(validationToken).split('+').join(' ')
+        return {
+          status: 200,
+          headers: {
+            'Content-Type': 'text/plain',
+          },
+          body: str,
+        }
+      }
+    }
+
+    if (!req.body) {
+      logger.warn('Handler received an empty body')
+      return
+    }
+
+    const body = JSON.parse(req.body)
+    if (!body.value) {
+      logger.warn('Missing property value in the body')
+      return
+    }
+
+    body.value.forEach(async (message: any) => {
```

Comment:
**issue:** please use the `for` operator

```ts
for (const message of body.value) { ... }
```
the will prevent floating promises rejections

also, I'd extract the content of the for loop in a dedicated function

Reply:
**chore:** instead of doing manual data validation, please use `zod` to validate the request payload only once and then you can assume that the data has the desired format.